services/api/kalvi/middlewares/AdminUserMiddleware.py

```diff
@@ -0,0 +1,30 @@
+# custom_middleware.py
+from rest_framework import status
+from django.http import JsonResponse
+from rest_framework_simplejwt.tokens import UntypedToken
+from db.models import User
+from rest_framework_simplejwt.exceptions import InvalidToken, TokenError
+
+class AdminUserMiddleware:
+    def __init__(self, get_response):
+        self.get_response = get_response
+    def __call__(self, request):
+        if 'HTTP_AUTHORIZATION' in request.META:
+            try:
+                # Extract token and retrieve user object with better error handling
+                token = UntypedToken(request.META['HTTP_AUTHORIZATION'].split()[1])
+                user_id = token.payload['user_id']
+            except (InvalidToken, TokenError):
+                return JsonResponse({'error': "Invalid or expired token"}, status=status.HTTP_401_UNAUTHORIZED)
+            try:
+                user = User.objects.get(pk=user_id)
```

Comment:
I was thinking, this will be a costly process ie., hitting users table for every single API. Instead, can we get this information from the JWT token? role=admin|learner. If this is not available, we we add this white creating a JWT token. ?

Reply:
jwt tokens doesn't provide it by default nor does request object (like request.user.is_superuser since we are not using default django authentication) , adding user role in tokens will be good idea but customizing it i dont know if it will change its default characteristics elsewhere like default backend token authentication.  i need to see those things before assuring you.